### PR TITLE
Linux 4.14, 4.19, 5.0+ compat: SIMD save/restore

### PR DESCRIPTION
**Pull request by behlendorf**
### Motivation and Context

Issue #9346

### Description

Contrary to initial testing we cannot rely on these kernels to
invalidate the per-cpu FPU state and restore the FPU registers.
Therefore, the kfpu_begin() and kfpu_end() functions have been
updated to unconditionally save and restore the FPU state.

### How Has This Been Tested?

Manually tested on a  4.19 and 5.2 kernel using the very effective
reproducer provided in #9346.  While scrubbing a pool to force a
large number of checksum operations execute `mprime -t` to run
the internal mprime stress tests.

Without this change `mprime` errors out almost immediately, with
this PR applied no errors have been observed for either kernel.
The xsave, fxsr, and fnsave save/restore call paths where each
tested by modifying the code to force each callpath, and then
limiting the allowed SIMD instruction as appropriate (e.g. sse-only
for fxsr).

@Fabian-Gruenbichler would you mind reviewing this change.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: include/os/linux/kernel/linux/simd_x86.h [hunk 1]
```diff
@@ -126,38 +126,64 @@
...
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				test_thread_flag(TIF_NEED_FPU_LOAD))
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				current->thread.fpu.initialized)
-#else
-#define	kfpu_allowed()		0
-#endif
+#define	kfpu_allowed()		(current->flags & PF_KTHREAD)
+#define	ex_handler_fprestore	ex_handler_default
```
Comment (Fabian-Gruenbichler) [inline include/os/linux/kernel/linux/simd_x86.h]: this single define might be enough to just allow use to call
copy_fpregs_to_kernel in kfpu_begin
copy_kernel_to_fpregs in kfpu_end (with missing exception handling, but that is missing from the current code in master and from this PR anyway ;))

it might be an easier-to-understand quick fix for master?

File: include/os/linux/kernel/linux/simd_x86.h [hunk 2]
```diff
@@ -126,38 +126,64 @@
...
+#define	kfpu_fnsave(addr)	__asm("fnsave %0" : "=m" (*(addr)))
+#define	kfpu_fxrstor(addr)	__asm("fxrstor %0" : : "m" (*(addr)))
+#define	kfpu_fxrstorq(addr)	__asm("fxrstorq %0" : : "m" (*(addr)))
+#define	kfpu_frstor(addr)	__asm("frstor %0" : : "m" (*(addr)))
 
 static inline void
 kfpu_initialize(void)
 {
 	WARN_ON_ONCE(!(current->flags & PF_KTHREAD));
+	current->thread.fpu.last_cpu = -1;
```
Comment (Fabian-Gruenbichler) [inline include/os/linux/kernel/linux/simd_x86.h]: this is `__fpu_invalidate_fpregs_state(&current->thread.fpu);` which would have made for an easier to read diff ;)

File: include/os/linux/kernel/linux/simd_x86.h [hunk 3]
```diff
@@ -126,38 +126,64 @@
...
+#define	kfpu_allowed()		(current->flags & PF_KTHREAD)
+#define	ex_handler_fprestore	ex_handler_default
+
+/*
+ * FPU save and restore instructions.
+ */
+#define	__asm			__asm__ __volatile__
+#define	kfpu_fxsave(addr)	__asm("fxsave %0" : "=m" (*(addr)))
+#define	kfpu_fxsaveq(addr)	__asm("fxsaveq %0" : "=m" (*(addr)))
+#define	kfpu_fnsave(addr)	__asm("fnsave %0" : "=m" (*(addr)))
```
Comment (Fabian-Gruenbichler) [inline include/os/linux/kernel/linux/simd_x86.h]: the kernel assembly has an additional fwait after fnsave, since fsave/fnsave clear the registers. not sure whether it is needed here as well?

File: include/os/linux/kernel/linux/simd_x86.h [hunk 4]
```diff
@@ -172,46 +198,76 @@ kfpu_begin(void)
...
-	} else if (use_fxsr()) {
-		error = copy_kernel_to_fxregs_err(&state->fxsave);
+	if (static_cpu_has(X86_FEATURE_XSAVE)) {
+		kfpu_restore_xsave(&current->thread.fpu.state.xsave, ~0);
+	} else if (static_cpu_has(X86_FEATURE_FXSR)) {
+		kfpu_restore_fxsr(&current->thread.fpu.state.fxsave);
 	} else {
-		error = copy_kernel_to_fregs_err(&state->fsave);
+		kfpu_restore_fsave(&current->thread.fpu.state.fsave);
 	}
```
Comment (Fabian-Gruenbichler) [inline include/os/linux/kernel/linux/simd_x86.h]: this part is still missing the error/exception handling, but that is probably out of scope for this PR..

File: include/os/linux/kernel/linux/simd_x86.h (hunk 1 above)
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: Unfortunately it's not since the expansion of the `kernel_insn` macro occurs in the static inline function.  That's probably for the best since doing the save/restore ourselves ensures any future kernel changes in this area won't effect this code.

File: include/os/linux/kernel/linux/simd_x86.h (hunk 2 above)
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: Right, in the end I wanted to avoid this dependency since it was easy to do so.  One less interface which could cause us trouble latter.

File: include/os/linux/kernel/linux/simd_x86.h (hunk 4 above)
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: That was my thinking, we can always revist this in a follow up PR.

File: include/os/linux/kernel/linux/simd_x86.h (hunk 3 above)
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: I can add it, just in case, but I wasn't sure why it would be needed.  I also noticed in the FreeBSD implementation it was not present.

File: include/os/linux/kernel/linux/simd_x86.h [hunk 5]
```diff
@@ -172,46 +198,76 @@ kfpu_begin(void)
...
+
+static inline void
+kfpu_restore_fsave(struct fregs_state *addr)
+{
+	kfpu_frstor(addr);
 }
 
 static inline void
 kfpu_end(void)
 {
```
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: I'll add some logic to handle this.

**Comment by behlendorf** [comment]
Updated to address @Fabian-Gruenbichler 's feedback.  Thanks!

**Comment by Fabian-Gruenbichler** [comment]
as discussed in #9346, this PR is apparently still incomplete/broken on some systems.. I'll try to dig in today - please don't merge it yet ;)

**Comment by behlendorf** [comment]
@Fabian-Gruenbichler any assistance would be appreciated, and nothing will be merged until we get this sorted out.  Unfortunately, I still haven't been able to reproduce the issue in my testing.  I saw in  #9346 you were able to reproduce the issue though, would you mind sharing your reproduction steps.

**Comment by Fabian-Gruenbichler** [comment]
### SETUP

Debian Buster, baremetal:

```
vendor_id       : GenuineIntel
cpu family      : 6
model           : 79
model name      : Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz
stepping        : 1
microcode       : 0xb00002e
cpu MHz         : 2099.815
cache size      : 20480 KB
fpu             : yes
fpu_exception   : yes
cpuid level     : 20
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips        : 4199.63
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
power management:
```

following FPU information is reported on startup:
```
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: Enabled xstate features 0x7, context size is 832 bytes, using 'standard' format.
```

RAIDZ-1 with 4 SSDs, default 0.8.2 settings

test consist of the following operations in parallel:

```
$ mprime -t
$ dd if=/dev/zero of=/testpool/testfile.dat bs=16M count=20000 oflag=direct status=progress
$ zpool scrub
```

the scrub gets repeated from time to time if it has finished. whole procedure left running until mprime reports errors, or a few test rounds have passed successfully.

I tested the following kernels:
```
/boot/vmlinuz-4.19.0-6-amd64 (stock Debian Buster)
/boot/vmlinuz-4.19.77-041977-generic (4.19.77 mainline, Ubuntu PPA)
/boot/vmlinuz-5.0.21-2-pve (our Proxmox VE kernel, based on Ubuntu Disco's)
/boot/vmlinux-5.2.0-050200-generic (5.2 mainline, Ubuntu PPA)
/boot/vmlinuz-5.2.19-050219-generic (5.2.19 mainline, Ubuntu PPA)
/boot/vmlinuz-5.3.5-050305-generic (5.3.5 mainline, Ubuntu PPA)
```

[Ubuntu Mainline PPA](http://kernel.ubuntu.com/~kernel-ppa/mainline/)

each individual kernel+code combination got tested with the following sequence:
```
git checkout FOO
git clean -xdf
./autogen.sh && ./configure && make -j $(nproc)
./scripts/zfs.sh
./cmd/zpool/zpool import testpool
TEST
./scripts/zfs.sh -u
```
`dmesg -wk` got monitored to ensure correct modules are loaded/unloaded

I tested both this PR, current master, and current master with the following diff applied:
```
commit 3f5676a332490b12c8ee4e4052861741288f55a1
Author: Thomas Lamprecht <t.lamprecht@proxmox.com>
Date:   Wed Sep 25 10:48:48 2019 +0200

    FPU register save/restore is also required on 5.0 kernels

    NOTE: the kernel needs to have the copy_kernel_to_xregs_err,
    copy_kernel_to_fxregs_err and copy_kernel_to_fregs_err functions
    backported for this to work.

    Signed-off-by: Thomas Lamprecht <t.lamprecht@proxmox.com>

diff --git a/include/os/linux/kernel/linux/simd_x86.h b/include/os/linux/kernel/linux/simd_x86.h
index c59ba4174..869c3909b 100644
--- a/include/os/linux/kernel/linux/simd_x86.h
+++ b/include/os/linux/kernel/linux/simd_x86.h
@@ -172,7 +172,6 @@ kfpu_begin(void)
        preempt_disable();
        local_irq_disable();

-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
        /*
         * The current FPU registers need to be preserved by kfpu_begin()
         * and restored by kfpu_end().  This is required because we can
@@ -181,11 +180,11 @@ kfpu_begin(void)
         * context switch.
         */
        copy_fpregs_to_fpstate(&current->thread.fpu);
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
+
+
+#if defined(HAVE_KERNEL_FPU_INITIALIZED)
        /*
-        * There is no need to preserve and restore the FPU registers.
-        * They will always be restored from the task's stored FPU state
-        * when switching contexts.
+        * Was removed with 5.2 as it was always set to 1 there
         */
        WARN_ON_ONCE(current->thread.fpu.initialized == 0);
 #endif
@@ -194,7 +193,6 @@ kfpu_begin(void)
 static inline void
 kfpu_end(void)
 {
-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
        union fpregs_state *state = &current->thread.fpu.state;
        int error;

@@ -206,7 +204,6 @@ kfpu_end(void)
                error = copy_kernel_to_fregs_err(&state->fsave);
        }
        WARN_ON_ONCE(error);
-#endif

        local_irq_enable();
        preempt_enable();
```

### RESULTS
*works* - no ill-effects noticed
*broken A* - mprime reports rounding errors with values around ~0.5, with < 0.4 expected
*broken B* - mprime reports rounding errors with values all over the place

sometimes the errors get reported within seconds, sometimes it takes a few minutes. I am very happy about the KABI check parallelization btw, without it this would have taken ages :-P

## PR
- 4.19.0-6-amd64: broken B
- 4.19.77-041977-generic: broken B
- 5.0.21-2-pve: broken B
- 5.2.0-050200-generic: broken B
- 5.2.19-050219-generic: broken B
- 5.3.5-050305-generic: broken B

=> very consistently broken

## master
- 4.19.0-6-amd64: broken A (expected)
- 4.19.77-041977-generic: broken A (expected)
- 5.0.21-2-pve: broken A (expected)
- 5.2.0-050200-generic: broken B
- 5.2.19-050219-generic: broken B
- 5.3.5-050305-generic: broken B

=> breakage switches from no save/restore < 5.2 to broken save/restore >= 5.2 ?

## patched master
- 4.19.0-6-amd64: not applicable (missing backport of _err functions)
- 4.19.77-041977-generic: (missing backport of _err functions)
- 5.0.21-2-pve: works (expected)
- 5.2.0-050200-generic: broken B (expected == master)
- 5.2.19-050219-generic: broken B (expected == master)
- 5.3.5-050305-generic: broken B (expected == master)

=> save restore works on < 5.2 (with backported _err restore), breaks on >= 5.2 ?

5.2 introduced the "defer FPU state load until return to userspace" change, which seems like a likely culprit for why (patched) master breaks then. but OTOH, this would mean that the code in master never worked for any kernel (< 5.2 was broken since it did no copy/restore whatsoever, and >= 5.2 did a broken copy/restore)?

I don't understand yet why the PR is broken across the board - it does make sense that it is broken on >= 5.2 since it does essentially the same as master (I started some tests yesterday, but those were all on 5.3 which I only realized late in the testing was broken for master as well, so most of that testing went into the wrong direction :-/)

**Comment by behlendorf** [comment]
@Fabian-Gruenbichler thank's for digging in to this.  I ended up coming to basically the same conclusion, we're somehow leaving the kernel FPU in an inconsistent state despite the save/restore to the task struct.  Given that, I decided to try a slightly different approach.  The updated PR now uses a per-cpu variable to save/restore the FPU state to ensure nothing can modify it.  As before the basic idea is to put everything back as we found it so what the kernel thinks is valid is still valid and there's no need to invalidate the register state.  This is still a WIP and needs some more work/cleanup, but my early results are encouraging using the `mprime + dd + scrub` test case.

**Comment by Fabian-Gruenbichler** [comment]
giving it a spin atm - the approach looks sound, initial testing as well! :)

should be more robust since we just need to watch for new save/restore instructions appearing - that is usually known quite a while before they show up in actual hardware so we should not be in for surprises hopefully ;)

File: include/os/linux/kernel/linux/simd_x86.h [hunk 6]
```diff
@@ -126,96 +124,191 @@
...
+			kfpu_fini();
+			return (ENOMEM);
+		}
+	}
+
+	return (0);
+}
+
+#define	kfpu_allowed()		1
+#define	ex_handler_fprestore	ex_handler_default
```
[the remark this answers is not in the document]
Comment (Fabian-Gruenbichler) [reply include/os/linux/kernel/linux/simd_x86.h]: we want to do our own error handling in the mid-term, so having all the save/restore assembly code here is a good idea anyway.

File: module/zcommon/zfs_prop.c [hunk 7]
```diff
@@ -865,17 +865,35 @@ zfs_prop_align_right(zfs_prop_t prop)
 #endif
 
 #if defined(_KERNEL)
+
+#if defined(HAVE_KERNEL_FPU_INTERNAL)
+#include <linux/simd.h>
+
+struct fpu **zfs_kfpu_fpregs;
+EXPORT_SYMBOL(zfs_kfpu_fpregs);
```
Comment (Fabian-Gruenbichler) [inline module/zcommon/zfs_prop.c]: same as above, the fpu struct is more than just the saved fpregs state..

File: include/os/linux/kernel/linux/simd_x86.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: I absolutely agree, though I'd prefer to tackle this in a follow up PR after things are solid.

File: include/os/linux/kernel/linux/simd_x86.h [hunk 8]
```diff
@@ -126,96 +124,191 @@
...
+}
+
+static inline void
+kfpu_restore_xsave(struct xregs_state *addr, uint64_t mask)
+{
+	uint32_t low, hi;
+
+	low = mask;
+	hi = mask >> 32;
+	XSTATE_XRESTORE(addr, low, hi);
```
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: See previous comment, follow up PR.

File: include/os/linux/kernel/linux/simd_x86.h [hunk 9]
```diff
@@ -126,96 +124,191 @@
...
-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
-	__fpu_invalidate_fpregs_state(&current->thread.fpu);
-	set_thread_flag(TIF_NEED_FPU_LOAD);
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
-	__fpu_invalidate_fpregs_state(&current->thread.fpu);
-	current->thread.fpu.initialized = 1;
-#endif
+	low = mask;
+	hi = mask >> 32;
+	XSTATE_XSAVE(addr, low, hi, err);
```
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: Good idea

File: include/os/linux/kernel/linux/simd_x86.h [hunk 10]
```diff
@@ -126,96 +124,191 @@
...
+	int cpu;
+
+	zfs_kfpu_fpregs = kzalloc(num_possible_cpus() *
+	    sizeof (struct fpu *), GFP_KERNEL);
+	if (zfs_kfpu_fpregs == NULL)
+		return (ENOMEM);
+
+	for_each_possible_cpu(cpu) {
+		zfs_kfpu_fpregs[cpu] = kmalloc_node(sizeof (struct fpu),
+		    GFP_KERNEL, cpu_to_node(cpu));
```
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: Good thought.  I've updated the patch to zero this memory when initially allocated.  As for the need to re-initialize after a restore since we're using dedicated memory for this only available to zfs I don't believe we need to re-initialize it after each restore.

File: include/os/linux/kernel/linux/simd_x86.h [hunk 11]
```diff
@@ -126,96 +124,191 @@
...
 /*
  * When the kernel_fpu_* symbols are unavailable then provide our own
- * versions which allow the FPU to be safely used in kernel threads.
- * In practice, this is not a significant restriction for ZFS since the
- * vast majority of SIMD operations are performed by the IO pipeline.
+ * versions which allow the FPU to be safely used.
  */
+#if defined(HAVE_KERNEL_FPU_INTERNAL)
+
+extern struct fpu **zfs_kfpu_fpregs;
```
[the remark this answers is not in the document]
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_x86.h]: Fixed, converted to `union fpregs_state`

File: module/zcommon/zfs_prop.c (hunk 7 above)
Comment (behlendorf) [reply module/zcommon/zfs_prop.c]: Fixed

**Comment by behlendorf** [comment]
This is ready for another round of review and testing.

* (Most) review feedback incorporated.
* Additional cleanup / refactoring
* Remove taskq dispatch code for encryption and micro benchmarks (no longer needed)

File: include/os/linux/kernel/linux/simd_x86.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Fabian-Gruenbichler) [reply include/os/linux/kernel/linux/simd_x86.h]: sorry for not being verbose enough - what I am worried about is

task A does something with FPU
ZFS saves state (does this always write the full fsave/fxsave/xsave struct?)
ZFS restores state
task A
kernel clears HW FPU state if needed
task B
ZFS saves state (same question?)
ZFS restore (if the two questions above are not a yes, then this might now restore partial state from task A)
task B

or

task A does something with FPU
ZFS saves state (same question)
ZFS restores state
ZFS saves state (same question)
ZFS restores state
task A (potentially sees (partial) state from ZFS FPU operations in HW registers)

it's probably just a question of reading the spec - I won't have time for that today unfortunately, but hopefully on Monday.

File: include/os/linux/kernel/linux/simd_x86.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Fabian-Gruenbichler) [reply include/os/linux/kernel/linux/simd_x86.h]: the mask inlining/constification could already be done without switching away from XSTATE_XSAVE?

**Comment by Fabian-Gruenbichler** [comment]
It might make sense to split the revert of the taskq dispatching into its own commit - it's not yet part of 0.8.x, would make for easier backporting of the rest ;)

Currently giving this another round of battering, will report back with results.

**Comment by behlendorf** [comment]
@Fabian-Gruenbichler were you able to perform any testing with the PR?  Any results?

**Comment by Fabian-Gruenbichler** [comment]
> @Fabian-Gruenbichler were you able to perform any testing with the PR? Any results?

sorry for missing posting an update - yes, I left our reproducer and some additional stress-testing running for quite a while without any visible issues.

besides the mask constification (which definitely falls into the 'nit' category) there was only the part about whether we need clearing / re-initialization on each begin/end cycle:

> sorry for not being verbose enough - what I am worried about is
> 
> task A does something with FPU
> ZFS saves state (does this always write the full fsave/fxsave/xsave struct?)
> ZFS restores state
> task A
> kernel clears HW FPU state if needed
> task B
> ZFS saves state (same question?)
> ZFS restore (if the two questions above are not a yes, then this might now restore partial state from task A)
> task B
> 
> or
> 
> task A does something with FPU
> ZFS saves state (same question)
> ZFS restores state
> ZFS saves state (same question)
> ZFS restores state
> task A (potentially sees (partial) state from ZFS FPU operations in HW registers)

not sure whether you saw that or not? I am not very fond of Github's review interface..

basically the question is - do we need to care about
- whether saving into memory can leave some partial prior state in the fpregs struct
- that could than be restored by the next restore
(my gut feeling says no)

and
- whether restore can just restore SOME registers
(here I am not so sure!)

both could leak ZFS or other data to the task that runs after ZFS.

**Comment by behlendorf** [comment]
Sorry I missed that comment!

> does this always write the full fsave/fxsave/xsave struct?

Good question.  Yes, that is my understanding.  As long as we set all of the bits in the xsave mask (which we do) we should be saving and restoring the registers exactly as they were and not leaking any ZFS state.  We could try to get fancier and only save/restore the registers we're going to use, but I wanted to try and keep things simple.

**Comment by behlendorf** [comment]
@Fabian-Gruenbichler if you're satisfied with the current PR, could you please approve it.  If not, then let's see if we can get any remaining concerns sorted out.

**Comment by interduo** [comment]
@tonyhutter Are You planning to release 0.8.3 version if this PR will go upstream?

**Comment by tonyhutter** [comment]
@interduo yes, we'll include it in the next release assuming we don't find any other issues.

**Comment by interduo** [comment]
@tonyhutter my question was a little bit different - will this PR could be a purpose for next release? 
Because this finally closes one of the major performance issues (https://github.com/zfsonlinux/zfs/issues/8836 and https://github.com/zfsonlinux/zfs/issues/8793)

**Comment by vorsich-tiger** [comment]
> ... assuming we don't find any other issues.
@Fabian-Gruenbichler 
@behlendorf 
I'd appreciate very much if you'd help to establish here the final and valid set of assumptions that this new solution is based on.
I'm asking this because while trying to understand the validity of the new solution, I constantly ran into the problem that I wondered what kind of implicit assumptions about kernel/scheduler behaviour can be made (and which cannot).
Specifically I'd like to get *reasoned* confirmation that the following *potential* loophole in the scheme does *not* exist:
Let me start with a short summary of how *I* understand the new approach.
Please correct me where I'm wrong:
1. zfs module (when FPU API is not available) in future does not intend to rely on any specific kernel fpu treatment behaviour
2. thus, zfs has a per cpu private FPU state save/restore buffer
3. when any zfs in kernel thread starts using the fpu, it saves the existing/current fpu state to the thread-current cpu specific zfs save buffer
4. when any zfs in kernel thread stops using the fpu, it restores the "original" fpu state of the thread-current cpu specific zfs save buffer
5. OTOH - the current kernels seems to track 1 fpu state per *thread* though

My worries now concern the following:
What if there exists any (by intention) supported kernel/scheduler (i.e. maybe an RT kernel) which preemts a zfs in kernel thread between 3. and 4. ?
There I'd expect a problem if the scheduler assigns that cpu which was active at 3. (but did *not*yet* arrive at 4., so it has effectively "corrupted" the fpu state) to a completely different thread - in that case the kernel's view would be assuming the fpu's state is identical to the saved state at 3., right ?
The kernel *might* decide to *not* restore the fpu state from the new thread's save buffer if that thread and the zfs thread alternate on that cpu, right ?
What assumption can be made that this does not happen ?
Is this by general kernel design rules ?

**Comment by Fabian-Gruenbichler** [comment]
On October 26, 2019 1:21 pm, vorsich-tiger wrote:
>> ... assuming we don't find any other issues.
> @Fabian-Gruenbichler 
> @behlendorf 
> I'd appreciate very much if you'd help to establish here the final and valid set of assumptions that this new solution is based on.
> I'm asking this because while trying to understand the validity of the new solution, I constantly ran into the problem that I wondered what kind of implicit assumptions about kernel/scheduler behaviour can be made (and which cannot).
> Specifically I'd like to get *reasoned* confirmation that the following *potential* loophole in the scheme does *not* exist:
> Let me start with a short summary of how *I* understand the new approach.
> Please correct me where I'm wrong:
> 1. zfs module (when FPU API is not available) in future does not intend to rely on any specific kernel fpu treatment behaviour
> 2. thus, zfs has a per cpu private FPU state save/restore buffer
> 3. when any zfs in kernel thread starts using the fpu, it saves the existing/current fpu state to the thread-current cpu specific zfs save buffer
> 4. when any zfs in kernel thread stops using the fpu, it restores the "original" fpu state of the thread-current cpu specific zfs save buffer
> 5. OTOH - the current kernels seems to track 1 fpu state per *thread* though
> 
> My worries now concern the following:
> What if there exists any (by intention) supported kernel/scheduler (i.e. maybe an RT kernel) which preemts a zfs in kernel thread between 3. and 4. ?
> There I'd expect a problem if the scheduler assigns that cpu which was active at 3. (but did *not*yet* arrive at 4., so it has effectively "corrupted" the fpu state) to a completely different thread - in that case the kernel's view would be assuming the fpu's state is identical to the saved state at 3., right ?
> The kernel *might* decide to *not* restore the fpu state from the new thread's save buffer if that thread and the zfs thread alternate on that cpu, right ?
> What assumption can be made that this does not happen ?
> Is this by general kernel design rules ?

preempts and interrupts are disabled between 3 and 4, just like for the 
regular kernel mechanism (otherwise none of this would work ;)).

**Comment by tonyhutter** [comment]
@interduo I would say that this patch is a contributing factor to wanting to put out a new release, but there are other reasons to create a 0.8.3 release as well (like misc bugfixes that have gone into master). I imagine we'll start putting together a 0.8.3 patch list in the upcoming weeks.

**Comment by interduo** [comment]
@tonyhutter is there any hope to see 0.8.3 in this month?

**Comment by tonyhutter** [comment]
@interduo probably not, since Thanksgiving is this week.  I was planning to talk to @behlendorf tomorrow and go over the potential patchlist though.  There's ~260 patches in master that aren't in the release branch, and we need to figure out which ones we want to include.

**Comment by interduo** [comment]
So maybe there is a need to do 0.9.x-rc1 release and then include all the patches?
I am always thinking why the release cycle of zfs is taking always so long - last was at Sep 26 2019.
This would give us opportunity to do some more testing at testing envirnoments before final release (...)

**Comment by tonyhutter** [comment]
FYI - we actually plan to skip 0.9.x and go straight to a 2.0 release in 2020.  2.0 will include FreeBSD support.  See slide 21 here:
https://drive.google.com/file/d/197jS8_MWtfdW2LyvIFnH58uUasHuNszz/view

The 2.0 release will include all the patches from master.  I image we'll do 2.0-rc releases like we did with 0.7.0 and 0.8.0.

**Comment by faern** [comment]
Ah, the good old Windows/iPhone trick of skipping version 9.

@interduo I think a few months is an extremely fast release cycle for a filesystem. Right now I'm also eager for SIMD fixes, but except for that I don't want what's taking care of all my data to change often at all.

**Comment by spacelama** [comment]
I've had some general unstableness since updating to 0.8.2-2 (debian backports).  Including general protection faults not apparently in ZFS code (alas, I've lost all the backtraces), and just hangs that went unlogged.  Debian do have the their own patch to the patch: 
zfs-linux (0.8.2-2) unstable; urgency=medium
  * Remove a patch implicated in data corruption (related to
    linux-5.0-simd-compat.patch)

Any chance they've screwed up the FPU restore, what would be the symptoms, and would they manifest at random places outside the ZFS code?

What's the patch they *should* be applying?

**Comment by cdluminate** [comment]
@spacelama What you mentioned in the changelog is exactly this patch: 

https://salsa.debian.org/zfsonlinux-team/zfs/blob/master/debian/patches/series#L14
https://salsa.debian.org/zfsonlinux-team/zfs/blob/master/debian/patches/Fix-CONFIG_X86_DEBUG_FPU-build-failure.patch

And I'd say that we are not patching ZFS without forwarding to upstream, except for those distro-specific ones.

**Comment by interduo** [comment]
@tonyhutter is there any hope to see 0.8.3 in december or january?

**Comment by bobobo1618** [comment]
Did this make it into the 0.8.3 release? I see in [the changes](https://github.com/zfsonlinux/zfs/releases/tag/zfs-0.8.3):
- SIMD: Use alloc_pages_node to force alignment (seems to be 35155c013240ce14860b43ebc4803c2a5eea78f8)
- Linux 5.0 compat: SIMD compatibility (seems to be 62c034f6d45df04fc81d6c7ca5bd884e17bfee19)

But I don't see 10fa254.

**Comment by bobobo1618** [comment]
Ah, I see. 62c034f squashes 10fa254 and a few other commits, so yes, this is in 0.8.3.
